File: 198. House Robber.java
class Solution {
    int ans = 0;
    Map<Integer, Integer> checked = new HashMap<>();
    public int rob(int[] nums) {
        for(int i = 0; i < nums.length; i ++) {
            find(nums, i, nums[i]);
        }
        return ans;
    }
    
    public void find(int[] nums, int step, int sum) {
        ans = Math.max(ans, sum);
        if(step >= nums.length) {            
            return;
        }
        if(checked.getOrDefault(step, 0) >= sum) {
            return;
        }
        checked.put(step, sum);
        
        for(int i = step + 2; i < nums.length; i++) {
            find(nums, i, sum + nums[i]);
        }
    }
}

File: 198. House Robber.kt
class Solution {
    fun rob(nums: IntArray): Int {
        val maxRob = IntArray(nums.size + 1) { 0 }
        val N = nums.size
        maxRob[N] = 0
        maxRob[N-1] = nums[N-1]

        for (i in N-2 downTo 0) {
            maxRob[i] = Math.max(maxRob[i+2] + nums[i], maxRob[i+1])
        }

       return maxRob[0]
    }
}
